File: haskell/P06Applicative.hs
import Data.Typeable
import Test.HUnit.Base

main :: IO ()
main = do
  show (typeOf fmapResult) @?= "Maybe (Int -> Int)"
  applyResult @?= Just 5

some2 = Just 2
some3 = Just 3

fmapResult = fmap' f some2

applyResult = apply' fmapResult some3

f :: Int -> Int -> Int
f x y = x + y

class Functor' f where
  fmap' :: (a -> b) -> f a -> f b

class Functor' f => Applicative' f where
  apply' :: f (a -> b) -> f a -> f b

instance Functor' Maybe where
  fmap' _ Nothing = Nothing
  fmap' f (Just a) = Just (f a)

instance Applicative' Maybe where
  apply' _ Nothing = Nothing
  apply' Nothing _ = Nothing
  apply' (Just g) (Just a) = Just (g a)

File: scala/src/P06Applicative.scala
object P06Applicative {
  def main(args: Array[String]): Unit = {
    val some2 = Some(2)
    val some3 = Some(3)

    val fmapResult = OptionFunctor fmap(f, some2)
    println(fmapResult)

    val applyResult = OptionApplicative apply(fmapResult, some3)
    assert(applyResult == Some(5))
  }

  val f: Int => Int => Int = x => y => x + y
}

trait Applicative[F[_]] extends Functor[F] {
  def apply[A, B](f: F[A => B], fa: F[A]): F[B]
}

object OptionApplicative extends Applicative[Option] {
  override def fmap[A, B](f: A => B, fa: Option[A]): Option[B] = {
    fa match {
      case None => None
      case Some(a) => Some(f(a))
    }
  }

  override def apply[A, B](f: Option[A => B], fa: Option[A]): Option[B] = {
    fa match {
      case None => None
      case Some(a) => f match {
        case None => None
        case Some(g) => Some(g(a))
      }
    }
  }
}